File: SpotifyTrendsApp.Server/Services/TokenService.cs
using System.Net.Http.Headers;
using System.Text;
using SpotifyTrendsApp.Server.Models;

namespace SpotifyTrendsApp.Server.Services;

public class TokenService : ITokenService
{
    private readonly HttpClient _httpClient;
    private readonly IConfiguration _configuration;
    private readonly ILogger<TokenService> _logger;
    private TokenInfo _currentToken;

    public TokenInfo CurrentToken
    {
        get => _currentToken;
        set => _currentToken = value;
    }

    public TokenService(HttpClient httpClient, IConfiguration configuration, ILogger<TokenService> logger)
    {
        _httpClient = httpClient;
        _configuration = configuration;
        _logger = logger;
        _currentToken = null!;
    }

    public async Task<TokenInfo> GetAccessTokenAsync(string code)
    {
        try
        {
            var clientId = _configuration["Spotify:ClientId"] ?? throw new InvalidOperationException("ClientId is not configured");
            var clientSecret = _configuration["Spotify:ClientSecret"] ?? throw new InvalidOperationException("ClientSecret is not configured");
            var redirectUri = _configuration["Spotify:RedirectUri"] ?? throw new InvalidOperationException("RedirectUri is not configured");
            var tokenEndpoint = _configuration["Spotify:TokenEndpoint"] ?? throw new InvalidOperationException("TokenEndpoint is not configured");

            _logger.LogInformation("Requesting access token for authorization code");

            var requestBody = new Dictionary<string, string>
            {
                { "grant_type", "authorization_code" },
                { "code", code },
                { "redirect_uri", redirectUri },

            };
            var authCredentials = Convert.ToBase64String(Encoding.UTF8.GetBytes($"{clientId}:{clientSecret}"));
            _httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Basic", authCredentials);
            _httpClient.DefaultRequestHeaders.Accept.Clear();
            _httpClient.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

            var requestContent = new FormUrlEncodedContent(requestBody);

            _logger.LogInformation("Sending request to Spotify API: {RequestBody}", requestBody);

            var response = await _httpClient.PostAsync(tokenEndpoint, requestContent);

            _logger.LogInformation("Received response from Spotify API: {StatusCode}", response.StatusCode);

            if (!response.IsSuccessStatusCode)
            {
                var errorContent = await response.Content.ReadAsStringAsync();
                _logger.LogError("Spotify API returned error: {ErrorContent}", errorContent);
            }

            response.EnsureSuccessStatusCode();

            var responseBody = await response.Content.ReadFromJsonAsync<AuthTokenResponse>() ?? throw new InvalidOperationException("Failed to deserialize OAuth token response");

            var token = new TokenInfo
            {
                AccessToken = responseBody.access_token!,  
                RefreshToken = responseBody.refresh_token ?? string.Empty,
                ExpiresAt = DateTime.UtcNow.AddSeconds(responseBody.expires_in)
            };

            _currentToken = token;
            return token;
        }
        catch (HttpRequestException ex)
        {
            _logger.LogError(ex, "Failed to obtain access token from Spotify");
            throw;
        }
    }

    public async Task<TokenInfo> RefreshAccessTokenAsync(string refreshToken)
    {
        try
        {
            var clientId = _configuration["Spotify:ClientId"] ?? throw new InvalidOperationException("ClientId is not configured");
            var clientSecret = _configuration["Spotify:ClientSecret"] ?? throw new InvalidOperationException("ClientSecret is not configured");
            var tokenEndpoint = _configuration["Spotify:TokenEndpoint"] ?? throw new InvalidOperationException("TokenEndpoint is not configured");

            _logger.LogInformation("Refreshing access token");

            var requestBody = new Dictionary<string, string>
        {
            { "grant_type", "refresh_token" },
            { "refresh_token", refreshToken },
        };

            var requestContent = new FormUrlEncodedContent(requestBody);

            var authCredentials = Convert.ToBase64String(Encoding.UTF8.GetBytes($"{clientId}:{clientSecret}"));
            _httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Basic", authCredentials);
            _httpClient.DefaultRequestHeaders.Accept.Clear();
            _httpClient.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

            var response = await _httpClient.PostAsync(tokenEndpoint, requestContent);
            response.EnsureSuccessStatusCode();

            var responseBody = await response.Content.ReadFromJsonAsync<AuthTokenResponse>() ?? throw new InvalidOperationException("Failed to deserialize OAuth token response");

            var token = new TokenInfo
            {
                AccessToken = responseBody.access_token!,  
                RefreshToken = responseBody.refresh_token ?? refreshToken,
                ExpiresAt = DateTime.UtcNow.AddSeconds(responseBody.expires_in)
            };

            _currentToken = token;
            _logger.LogInformation("Successfully refreshed access token");
            return token;
        }
        catch (HttpRequestException ex)
        {
            _logger.LogError(ex, "Failed to refresh access token from Spotify");
            throw;
        }
    }

    public async Task<string?> GetStoredRefreshTokenAsync()
        => await Task.FromResult(_currentToken?.RefreshToken);

    public async Task<string?> GetStoredAccessTokenAsync()
        => await Task.FromResult(_currentToken?.AccessToken);
}
public class AuthTokenResponse
{
    public string? access_token { get; set; }
    public string? token_type { get; set; }
    public int expires_in { get; set; }
    public string? refresh_token { get; set; }
    public string? scope { get; set; }
}

File: go-services/internal/server/service/token_service.go
package service

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"

	"github.com/sirupsen/logrus"

	"github.com/conor27nash/ListeningTrendsApp/go-services/internal/server/model"
	"github.com/conor27nash/ListeningTrendsApp/go-services/pkg/config"
)

// TokenService handles Spotify OAuth token operations
type TokenService struct {
	config     *config.Config
	httpClient *http.Client
	logger     *logrus.Logger
}

// NewTokenService creates a new TokenService instance
func NewTokenService(cfg *config.Config, logger *logrus.Logger) *TokenService {
	return &TokenService{
		config:     cfg,
		httpClient: &http.Client{},
		logger:     logger,
	}
}

// GetAccessToken exchanges an authorization code for access and refresh tokens
func (s *TokenService) GetAccessToken(code string) (*model.TokenInfo, error) {
	// Prepare form data
	data := url.Values{}
	data.Set("grant_type", "authorization_code")
	data.Set("code", code)
	data.Set("redirect_uri", s.config.RedirectURI)
	data.Set("client_id", s.config.SpotifyClientID)
	data.Set("client_secret", s.config.SpotifyClientSecret)

	// Create request
	req, err := http.NewRequest("POST", "https://accounts.spotify.com/api/token", strings.NewReader(data.Encode()))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")

	// Send request
	s.logger.Debug("Exchanging authorization code for tokens")
	resp, err := s.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	// Read response
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("token exchange failed with status %d: %s", resp.StatusCode, string(body))
	}

	// Parse response
	var tokenResponse model.OAuthTokenResponse
	if err := json.Unmarshal(body, &tokenResponse); err != nil {
		return nil, fmt.Errorf("failed to parse token response: %w", err)
	}

	s.logger.Info("Successfully exchanged code for tokens")

	return &model.TokenInfo{
		AccessToken:  tokenResponse.AccessToken,
		RefreshToken: tokenResponse.RefreshToken,
		ExpiresIn:    tokenResponse.ExpiresIn,
		TokenType:    tokenResponse.TokenType,
		Scope:        tokenResponse.Scope,
	}, nil
}

// RefreshAccessToken refreshes the access token using a refresh token
func (s *TokenService) RefreshAccessToken(refreshToken string) (*model.TokenInfo, error) {
	// Prepare form data
	data := url.Values{}
	data.Set("grant_type", "refresh_token")
	data.Set("refresh_token", refreshToken)
	data.Set("client_id", s.config.SpotifyClientID)
	data.Set("client_secret", s.config.SpotifyClientSecret)

	// Create request
	req, err := http.NewRequest("POST", "https://accounts.spotify.com/api/token", strings.NewReader(data.Encode()))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")

	// Send request
	s.logger.Debug("Refreshing access token")
	resp, err := s.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	// Read response
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("token refresh failed with status %d: %s", resp.StatusCode, string(body))
	}

	// Parse response
	var tokenResponse model.OAuthTokenResponse
	if err := json.Unmarshal(body, &tokenResponse); err != nil {
		return nil, fmt.Errorf("failed to parse token response: %w", err)
	}

	s.logger.Info("Successfully refreshed access token")

	// Refresh token might not be returned, so use the old one if not present
	if tokenResponse.RefreshToken == "" {
		tokenResponse.RefreshToken = refreshToken
	}

	return &model.TokenInfo{
		AccessToken:  tokenResponse.AccessToken,
		RefreshToken: tokenResponse.RefreshToken,
		ExpiresIn:    tokenResponse.ExpiresIn,
		TokenType:    tokenResponse.TokenType,
		Scope:        tokenResponse.Scope,
	}, nil
}
